archaius2-core/src/main/java/com/netflix/archaius/config/DependentConfigListener.java

```diff
@@ -0,0 +1,65 @@
+package com.netflix.archaius.config;
+
+import com.netflix.archaius.api.Config;
+import com.netflix.archaius.api.ConfigListener;
+
+import java.lang.ref.Reference;
+import java.lang.ref.WeakReference;
+import java.util.Optional;
+
+/**
+ * ConfigListener for the dependent/wrapper config paradigm. Most notably makes the reference to the dependent config
+ * a weak reference and removes this listener from the source config so that any abandoned dependent configs can be
+ * properly garbage collected.
+ *
+ * @param <T> The type of the dependent config
+ */
+public abstract class DependentConfigListener<T extends AbstractConfig> implements ConfigListener {
+    private final Reference<T> dependentConfigRef;
```

Comment:
Let's make this class package-private, it's really just an implementation detail for the other config classes.

Reply:
Done